### PR TITLE
[FSDP2] Fix backward when loss depends on non-output tensor

### PR DESCRIPTION
**Pull request by mikevsky**
Fixes https://github.com/pytorch/pytorch/issues/173709

## Summary
- When `reshard_after_forward=True`, `_register_pre_backward_hook` only hooks on module output tensors. If the loss depends on a cached intermediate instead of the output, hooks never fire and parameters stay resharded (0-size storage) during backward.
- Use `saved_tensors_hooks` during the module forward so that when any saved tensor is accessed during backward, parameters are unsharded first. `_pre_backward` is idempotent so the existing output-based hooks still work without conflict.

## Test plan
- Added `TestFullyShard1DTrainingBackward.test_backward_reshard_with_non_output_loss`
- Verified fix on PyTorch 2.9.0 container with `torchrun --nproc_per_node=2`

### CODE REVIEW

File: test/distributed/_composable/fsdp/test_fully_shard_training.py [hunk 1]
```diff
@@ -684,6 +684,44 @@ def step_post_hook(
...
+
+        torch.manual_seed(42)
+        model = Model().to(device_type)
+        fully_shard(model, reshard_after_forward=True)
+        x = torch.randn(4, 64, device=device_type.type, requires_grad=True)
+        _output = model(x)
+        loss = model.cached_intermediate.sum()
+        # Should not raise "storage of size 0" error
+        loss.backward()
+        for param in model.parameters():
```
Comment (weifengpy) [inline test/distributed/_composable/fsdp/test_fully_shard_training.py:720]: do you mean model.linear1.parameters()? otherwise linear2 grad are None and will raise assertion error

is this for review or just showcase the idea?
